grails-app/controllers/org/pih/warehouse/api/ProductSupplierApiController.groovy

```diff
@@ -28,4 +30,17 @@ class ProductSupplierApiController {
         productSupplierService.delete(params.id)
         render status: 204
     }
+
+    def create(ProductSupplierDetailsCommand productSupplierDetailsCommand) {
+        ProductSupplier productSupplier = productSupplierService.saveDetails(productSupplierDetailsCommand)
+
+        response.status = HttpStatus.CREATED.value()
+        render([data: productSupplier.toJson()] as JSON)
```

Comment:
You'll also need to add one for ProductSupplierPreference.
```
        JSON.registerObjectMarshaller(ProductSupplierPreference) { ProductSupplierPreference productSupplierPreference ->
            return productSupplierPreference.toJson()
        }
```


And the supplier object in the ProductSupplier.toJson() is missing an ID, which initially prevented me from copying the JSON returned by a GET request and sending it in a PUT request.

Reply:
Here's my attempt at the create endpoint. The validation exception should have a better message, but I hate that I need to pass a message there.
```
    def create(ProductSupplierDetailsCommand command) {
        if (command.hasErrors()) {
            throw new ValidationException("validation errors", command.errors)
        }

        ProductSupplier productSupplier = productSupplierService.saveProductSupplier(params.id, command)

        response.status = HttpStatus.CREATED.value()

        render([data: productSupplier] as JSON)
    }
```

Reply:
The reason I would prefer to use saveProductSupplier over saveDetails or save is because the ProductSupplierService will probably also house the code to save other objects like preferences and packages that are associated with a product supplier.

Reply:
Is there a deeper reason why you prefer Bootstrap over `toJson()` here? I think recently we've discussed that we wanted to avoid using Bootstrap and stick to `toJson()` for now even if it's not the best approach possible.

Reply:
The reason I went with `saveDetails` is for this method not to be confusing - doing it as `saveProductSupplier` in my opinion suggests, that we could update all product supplier's properties there - this is why I haven't named it like this.
But if you guys are ok with that, then I can change it.

Reply:
Regarding your first comment, we should continue to use the JSON marshaller framework (including marshaller registration in BootStrap.groovy) until we have migrated to JSON views.

Reply:
Regarding the second one, I understand your point. We just have a convention of using the name action<DomainClass>, like saveProduct, saveProductSupplier. 

We haven't really discussed having method names that imply scope. And I would say that the scope should be explicitly defined through better security measures, not though naming. But if we need 

If we want to adopt a new naming convention we should create a proposal on the new naming convention and agree on it before applying it. 

Here's an example of what a naming convention might look like. I don't like all of these convention, but I like the general idea of defining each convention, providing some justification and demonstration through a code example.
https://sergiodelamo.com/blog/grails-extra-conventions.html